dm/dm/master/openapi_controller_test.go

```diff
@@ -0,0 +1,376 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// MVC for dm-master's openapi server
+// Model(data in etcd): source of truth
+// View(openapi_view): do some inner work such as validate, filter, prepare parameters/response and call controller to update model.
+// Controller(openapi_controller): call model func to update data.
+
+package master
+
+import (
+	"context"
+	"testing"
+
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/tiflow/dm/checker"
+	"github.com/pingcap/tiflow/dm/dm/config"
+	"github.com/pingcap/tiflow/dm/dm/master/scheduler"
+	"github.com/pingcap/tiflow/dm/dm/pb"
+	"github.com/pingcap/tiflow/dm/openapi"
+	"github.com/pingcap/tiflow/dm/openapi/fixtures"
+	"github.com/pingcap/tiflow/dm/pkg/log"
+	"github.com/pingcap/tiflow/dm/pkg/terror"
+	"github.com/stretchr/testify/suite"
+)
+
+type OpenAPIControllerSuite struct {
+	suite.Suite
+
+	testSource *openapi.Source
+	testTask   *openapi.Task
+}
+
+func (s *OpenAPIControllerSuite) SetupSuite() {
```

Comment:
some branch(not some error branch which maybe hard to mock) is not covered, such as `startTask with removeMeta`, `force deleteTask`.

not sure if we should add in this pr(large enough), maybe later.

Reply:
this case is hard to mock, so i added in it-test now, see [here](https://github.com/Ehco1996/ticdc/blob/11d3081d3f6e32c713b8a9994a0d21a1efb51db1/dm/tests/openapi/run.sh#L290)

other comments addressed in https://github.com/pingcap/tiflow/pull/4627/commits/9991f20ede059430e4edbb460c7e451c64700c9d